packages/teams-js/test/internal/communication.spec.ts

```diff
@@ -1,67 +1,378 @@
-    const initPromise = communication.initializeCommunication(undefined);
-    await expect(initPromise).rejects.toThrowError('Initialization Failed. No Parent window found.');
-  });
-  it('processMessage fail if message has a missing data property', () => {
-    const event = { badData: '' } as any as DOMMessageEvent;
-    const result = communication.processMessage(event);
-    expect(result).toBeUndefined();
-  });
-  it('processMessage fail if message is empty', () => {
-    const event = {} as any as DOMMessageEvent;
-    const result = communication.processMessage(event);
-    expect(result).toBeUndefined();
-  });
-  it('processMessage fail if data property is not an object', () => {
-    const event = { data: '' } as any as DOMMessageEvent;
-    const result = communication.processMessage(event);
-    expect(result).toBeUndefined();
-  });
-  it('processMessage fail if message has random data', () => {
-    const event = { badData: '', notAnOrigin: 'blah' } as any as DOMMessageEvent;
-    const result = communication.processMessage(event);
-    expect(result).toBeUndefined();
-  });
-  it('processMessage fail if data is undefined', () => {
-    const event = { data: undefined } as any as DOMMessageEvent;
-    const result = communication.processMessage(event);
-    expect(result).toBeUndefined();
-  });
-  it('processMessage fail if data is null', () => {
-    const event = { data: null } as any as DOMMessageEvent;
-    const result = communication.processMessage(event);
-    expect(result).toBeUndefined();
-  });
-  it('shouldProcessMessage fail if message source is same window ', () => {
-    communication.Communication.currentWindow = window;
-    // window object should now equal Communication.currentWindow
-    const result = communication.shouldProcessMessage(window, 'testOrigin.com');
-    expect(result).toBe(false);
+      beforeEach(() => {
+        // Set a mock window for testing
+        utils = new Utils();
+        utils.mockWindow.parent = undefined;
+        app._initialize(utils.mockWindow);
+        communication.Communication.parentWindow = undefined;
+        GlobalVars.isFramelessWindow = false;
+      });
+
+      afterAll(() => {
+        communication.uninitializeCommunication();
+        GlobalVars.isFramelessWindow = false;
+      });
+
+      it('should throw if there is no parent window and no native interface on the current window', async () => {
+        app._initialize(undefined);
+        const initPromise = communication.initializeCommunication(undefined);
+        await expect(initPromise).rejects.toThrowError('Initialization Failed. No Parent window found.');
+      });
+
+      it('should receive valid initialize response from parent', async () => {
+        const initPromise = communication.initializeCommunication(undefined);
+        const initMessage = utils.findInitializeMessageOrThrow();
+
+        utils.respondToNativeMessage(initMessage, false, FrameContexts.content);
```

Comment:
>utils.respondToNativeMessage(initMessage, false, FrameContexts.content); [](http://example.com/codeflow?start=8&length=72)

I'm just going to ask questions and probably most of them are thing you can just tuck away for future refactoring. Some of them may just be dumb questions.

If `isFramelessWindow` is false, shouldn't `respondToNativeMessage` do nothing? Or throw? Or something? No native messages should be possible to receive or respond to in the frameless case?

Reply:
I'm guessing that something in the act of running the test changes `GlobalVars.isFramelessWindow` to be true?

Reply:
Your guess is correct. Anytime communication.initializeCommunication is called, there is no parent or opener of currentWindow, and the window has nativeInterface defined, then initializeCommunication sets isFramelessWindow to true. In this case, all of those things are true so the bit is flipped and we are a frameless window now.